src/fixate/core/switching.py

```diff
@@ -0,0 +1,720 @@
+        raise NotImplementedError
+
+
+def bit_generator() -> Generator[int, None, None]:
+    """b1, b10, b100, b1000, ..."""
+    return (1 << counter for counter in itertools.count())
+
+
+class PinValueAddressHandler(AddressHandler):
+    """Maps pins to bit values then combines the bit values for an update"""
+
+    def __init__(self):
+        super().__init__()
+        self._pin_lookup = {pin: bit for pin, bit in zip(self.pin_list, bit_generator())}
+
+    def set_pins(self, pins: Collection[Pin]):
+        value = sum(self._pin_lookup[pin] for pin in pins)
+        self._update_output(value)
+
+    def _update_output(self, value: int):
+        # perhaps it's easy to compose by passing the output
+        # function into __init__, like what we did with the VirtualMux?
+        bits = len(self.pin_list)
+        print(f"0b{value:0{bits}b}")
+
+
+class FTDIAddressHandler(PinValueAddressHandler):
+    """
+    An address handler which uses the ftdi driver to control pins.
+
+    We create this concrete address handler because we use it most
+    often. FT232 is used to bit-bang to shift register that are control
+    the switching in a jig.
+    """
+
+
+
+
+
+class VirtualAddressMap:
+    """
+    The supervisor loops through the attached virtual multiplexers each time a mux update is triggered.
+    """
+    def __init__(self, handlers: Sequence[AddressHandler]):
+        self._pin_to_handler: dict[Pin, AddressHandler] = {}
+
+        for handler in handlers:
+            self._pin_to_handler.update({pin: handler for pin in handler.pin_list})
+
+        self._pending_updates: list[PinUpdate] = []
+
+    def add_update(
+            self,
+            pin_update: PinUpdate,
+            trigger_update: bool = True
+    ) -> None:
+        """This method should be registered with each virtual mux to route pin changes."""
+        self._pending_updates.append(pin_update)
+
+        if trigger_update:
+            self._do_pending_updates()
+
+    def _do_pending_updates(self) -> None:
+        """
+        Collate pending updates and send pins to respective address handlers.
+
+        1.  For each pending update, combine pins to clear and pins to set.
+        2.  Find the longest `minimum_change_time` of all the updates
+        3.  Find the AddressHandler required to set each pin.
+        4.  Check if there is actually anything to change. Only write out
+            changes that are required.
+        5.  Do the setup phase
+        6.  Wait the required change time
+        7.  Do the final phase
+        """
+
+
+    # Not used in scripts
+    # @property
+    # def pin_values(self):
+    #     return list(
+    #         zip(
+    #             self.virtual_pin_list,
+    #             bits(
+    #                 self._virtual_pin_values_active,
+    #                 num_bits=len(self.virtual_pin_list),
+    #                 order="LSB",
+    #             ),
+    #         )
+    #     )
+
+    def active_pins(self) -> None:
```

Comment:
@jcollins1983 said
> I'm not sure what your plan is for the actie_pins method, but I use it when fault finding and debugging, so would be good to keep or at least have an equivalent.

Reply:
That is good feedback to have. Absolutely will keep anything that is useful, or replace with something better. When all the basics work, I'll have to go back and note everything that has been removed to double check nothing important has been deleted.

Reply:
I've ended up with `active_pins` on the address map and `active_signals` on the mux group
![image](https://github.com/PyFixate/Fixate/assets/7078979/23dbfe29-51b9-44b2-ad18-216d20d0fe4a)

I'm open to suggestions on how what the return value is, but I think keeping having those two distinct functions is useful